src/txmempool.h

```diff
@@ -816,6 +816,40 @@ class CTxMemPool
         assert(m_epoch.guarded()); // verify guard even when it==nullopt
         return !it || visited(*it);
     }
+
+    class CTxMemPoolChangeSet {
+    public:
+        explicit CTxMemPoolChangeSet(CTxMemPool* pool) : m_pool(pool) {}
+        ~CTxMemPoolChangeSet() = default;
+
+        CTxMemPoolChangeSet(const CTxMemPoolChangeSet&) = delete;
+        CTxMemPoolChangeSet& operator=(const CTxMemPoolChangeSet&) = delete;
+
+        using TxHandle = CTxMemPool::txiter;
+
+        TxHandle StageAddition(const CTransactionRef& tx, const CAmount fee, int64_t time, unsigned int entry_height, uint64_t entry_sequence, bool spends_coinbase, int64_t sigops_cost, LockPoints lp);
+        void StageRemoval(CTxMemPool::txiter it) { m_to_remove.insert(it); }
+
+        util::Result<CTxMemPool::setEntries> CalculateMemPoolAncestors(TxHandle tx, const Limits& limits) {
+            LOCK(m_pool->cs);
+            auto ret = m_pool->CalculateMemPoolAncestors(*tx, limits);
+            if (ret) m_ancestors.insert({tx, *ret});
+            else m_ancestors.erase(tx);
+            return ret;
+        }
+
+        void Apply() EXCLUSIVE_LOCKS_REQUIRED(cs_main);
+
+    private:
+        CTxMemPool* m_pool;
+        CTxMemPool::indexed_transaction_set m_to_add;
+        std::vector<CTxMemPool::txiter> m_entry_vec; // track the added transactions' insertion order
```

Comment:
My intention was to allow calling `StageAddition` and `StageRemoval` in any order.  I can add comments to that effect.

Reply:
Added docs in 7bc0f1f48c079b019f1d7a3fef00ae711f25d624

Reply:
Update: I realized that at least for now, there is a requirement that any added transactions are added in a topological order, with parents coming before children -- this is so that we can add things to the mempool's multi_index at `Apply()` time without creating any missing dependencies.  But removals can happen in any order and can be interleaved with the additions.